### PR TITLE
feat(merkle-tree): proof extraction helper

### PR DESCRIPTION
**Pull request by mrain**
closes: #XXXX

### This PR: 
Add a helper function to extract all leaves with proofs.

### Key places to review:

 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (main)
- [ ] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests
- [ ] Updated relevant documentation in the code
- [ ] Added relevant changelog entries to the `CHANGELOG.md` of touched crates.
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer

### CODE REVIEW

File: merkle_tree/src/macros.rs [hunk 1]
```diff
@@ -124,6 +124,20 @@ macro_rules! impl_merkle_tree_scheme {
...
+        impl<E, H, I, const ARITY: usize, T> $name<E, H, I, ARITY, T>
+        where
+            E: $crate::Element,
+            H: $crate::DigestAlgorithm<E, I, T>,
+            I: $crate::Index + $crate::ToTraversalPath<ARITY>,
+            T: $crate::NodeValue,
+        {
+            /// A helper function to collect all leaves with their corresponding proofs.
+            pub fn collect_leaves_with_proof(&self) -> $crate::Vec<(&I, &E, $crate::internal::MerkleTreeProof<T>)> {
+                let mut collector = $crate::Vec::new();
```
[the remark this answers is not in the document]
Comment (mrain) [reply merkle_tree/src/macros.rs]: in af8ae34

File: merkle_tree/src/internal.rs [hunk 2]
```diff
@@ -856,6 +856,37 @@ where
...
+                    current_proof.pop();
+                }
+            },
+            MerkleNode::Leaf {
+                elem,
+                value: _,
+                pos,
+            } => {
+                let mut proof = current_proof.clone();
+                proof.reverse();
```
[the remark this answers is not in the document]
Comment (mrain) [reply merkle_tree/src/internal.rs]: [af8ae34](https://github.com/EspressoSystems/jellyfish/pull/880/commits/af8ae34908f261dece3c09d25ade3912e6c2d60f)

**Comment by philippecamacho** [comment]
@mrain Added some extra test to handle edge cases in 096e8d46.

File: merkle_tree/src/append_only.rs [hunk 3]
```diff
@@ -361,6 +361,69 @@ mod mt_tests {
         );
     }
 
+    #[test]
+    fn test_mt_collect_leaves_with_proof() {
```
Comment (philippecamacho) [inline merkle_tree/src/append_only.rs]:
```suggestion
    fn test_mt_collect_leaves_with_proofs() {
```

File: merkle_tree/src/append_only.rs [hunk 4]
```diff
@@ -361,6 +361,69 @@ mod mt_tests {
...
+        test_mt_collect_leaves_with_proof_helper::<Fr254>();
+        test_mt_collect_leaves_with_proof_helper::<Fr377>();
+        test_mt_collect_leaves_with_proof_helper::<Fr381>();
+    }
+
+    fn test_mt_collect_leaves_with_proof_helper<F: RescueParameter>() {
```
Comment (philippecamacho) [inline merkle_tree/src/append_only.rs]:
```suggestion
    fn test_mt_collect_leaves_with_proofs_helper<F: RescueParameter>() {
```

File: merkle_tree/src/append_only.rs [hunk 5]
```diff
@@ -361,6 +361,69 @@ mod mt_tests {
...
+                *pos
+            );
+            let (lookup_elem, lookup_proof) = mt.lookup(*pos).expect_ok().unwrap();
+            assert_eq!(*elem, lookup_elem);
+            assert_eq!(*proof, lookup_proof);
+        }
+    }
+
+    #[test]
+    fn test_mt_collect_leaves_with_proof_edge_cases() {
```
Comment (philippecamacho) [inline merkle_tree/src/append_only.rs]:
```suggestion
    fn test_mt_collect_leaves_with_proofs_edge_cases() {
```

File: merkle_tree/src/append_only.rs [hunk 6]
```diff
@@ -361,6 +361,69 @@ mod mt_tests {
...
+        test_mt_collect_leaves_with_proof_edge_cases_helper::<Fr254>();
+        test_mt_collect_leaves_with_proof_edge_cases_helper::<Fr377>();
+        test_mt_collect_leaves_with_proof_edge_cases_helper::<Fr381>();
+    }
+
+    fn test_mt_collect_leaves_with_proof_edge_cases_helper<F: RescueParameter>() {
```
Comment (philippecamacho) [inline merkle_tree/src/append_only.rs]:
```suggestion
    fn test_mt_collect_leaves_with_proofs_edge_cases_helper<F: RescueParameter>() {
```

File: merkle_tree/src/internal.rs [hunk 7]
```diff
@@ -856,6 +856,33 @@ where
             MerkleNode::ForgottenSubtree { .. } => Err(MerkleTreeError::ForgottenLeaf),
         }
     }
+
+    /// Extract all elements with proof
+    pub(crate) fn collect_all_with_proof<'a>(
```
Comment (philippecamacho) [inline merkle_tree/src/internal.rs]:
```suggestion
    pub(crate) fn collect_all_with_proofs<'a>(
```

**Review by philippecamacho — DISMISSED** [review]
Added some extra test. 

Nit: some renaming suggestion.

**Comment by mrain** [comment]
thanks @philippecamacho , renaming is clean right now

**Review by philippecamacho — APPROVED** [review]
LGTM
